relay-cogs/src/lib.rs

```diff
@@ -0,0 +1,189 @@
+            Self::UnattributedMetrics => "unattributed_metrics",
+            Self::UnattributedEnvelope => "unattributed_envelope",
+            Self::Transactions => "transactions",
+            Self::Errors => "errors",
+            Self::Spans => "spans",
+            Self::Sessions => "sessions",
+            Self::ClientReports => "client_reports",
+            Self::CheckIns => "check_ins",
+            Self::Replays => "replays",
+            Self::MetricMeta => "metric_meta",
+            Self::MetricsTransactions => "metrics_transactions",
+            Self::MetricsSpans => "metrics_spans",
+            Self::MetricsSessions => "metrics_sessions",
+            Self::MetricsCustom => "metrics_cusomt",
+            Self::MetricsUnsupported => "metrics_unsupported",
+        }
+    }
+}
+
+impl utils::Enum<16> for AppFeature {
+    fn from_index(index: usize) -> Option<Self> {
+        let r = match index {
+            0 => Self::Unattributed,
+            1 => Self::UnattributedMetrics,
+            2 => Self::UnattributedEnvelope,
+            3 => Self::Transactions,
+            4 => Self::Errors,
+            5 => Self::Spans,
+            6 => Self::Sessions,
+            7 => Self::ClientReports,
+            8 => Self::CheckIns,
+            9 => Self::Replays,
+            10 => Self::MetricMeta,
+            11 => Self::MetricsTransactions,
+            12 => Self::MetricsSpans,
+            13 => Self::MetricsSessions,
+            14 => Self::MetricsCustom,
+            15 => Self::MetricsUnsupported,
+            _ => return None,
+        };
+
+        Some(r)
+    }
+
+    fn to_index(value: Self) -> usize {
+        match value {
+            Self::Unattributed => 0,
+            Self::UnattributedMetrics => 1,
+            Self::UnattributedEnvelope => 2,
+            Self::Transactions => 3,
+            Self::Errors => 4,
+            Self::Spans => 5,
+            Self::Sessions => 6,
+            Self::ClientReports => 7,
+            Self::CheckIns => 8,
+            Self::Replays => 9,
+            Self::MetricMeta => 10,
+            Self::MetricsTransactions => 11,
+            Self::MetricsSpans => 12,
+            Self::MetricsSessions => 13,
+            Self::MetricsCustom => 14,
+            Self::MetricsUnsupported => 15,
+        }
+    }
+}
+
+// A COGS measurement.
+#[derive(Debug, Clone, Copy)]
+pub struct CogsMeasurement {
+    /// The measured resource.
+    pub resource: ResourceId,
+    /// The measured app feature.
+    pub feature: AppFeature,
+    /// The measurement value.
+    pub value: Value,
+}
+
+/// A COGS measurement value.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Value {
+    /// A time measurement.
+    Time(Duration),
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::utils::Enum;
+
+    use super::*;
+
+    #[test]
+    fn test_app_feature() {
+        for i in 0.. {
+            let Some(f) = AppFeature::from_index(i) else {
```

Comment:
Does this test only verify indexes in `from_index` -> `to_index`? Should we also verify the other way around?

Reply:
It implicitly does, if `from_index` would yield the same variant twice, the assert below with `to_index` would at least fail once.